Remove SubProcess related code from the DependencyGraph service

wddgit opened the pull request:
#### PR description:

Remove the SubProcess related code from the DependencyGraph service. Given that the ability to run SubProcess has already been removed, this should not affect the output of anything.

There was also a minor bug in the limit when removing vertex entries corresponding to deleted modules. The limit should have been 0 instead of 1, although the likelihood of the first entry needing to be erased is small (not sure what the consequence would be even if it was not erased, maybe nothing significant.

Also checked the header includes and made some additions and subtractions.

#### PR validation:

Relies on existing tests. This does not add new functionality. Most just deletions of dead code.

wddgit committed:
Remove SubProcess related code from the DependencyGraph service

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"2f75a2238bb3efc580fed6680b5df10bc7635d5e":{"files":["FWCore/Services/plugins/DependencyGraph.cc"],"squashed":false,"time":1746029501}},"emoji":{},"last_seen_sha":"2f75a2238bb3efc580fed6680b5df10bc7635d5e","signatures":{}} -->

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47996/44664

wddgit commented:
please test

wddgit commented:
abort test

Just noticed something is wrong with this still. Seems to interact badly with module deletion... Debugging.

fwyzard commented on FWCore/Services/plugins/DependencyGraph.cc:
```diff
@@ -398,7 +373,7 @@ void DependencyGraph::lookupInitializationComplete(PathsAndConsumesOfModulesBase
 
   // remove the nodes corresponding to the modules that have been removed from the process
-  for (int i = boost::num_vertices(m_graph) - 1; i > 1; --i) {
+  for (int i = boost::num_vertices(m_graph) - 1; i > 0; --i) {
```
IIRC `0` should be the `source`, so it cannot be removed ?

makortel replied:
I don't think we guarantee `source` to have module ID 0, i.e. if that happens today, it is accidental and may change in the future. Therefore looping over all nodes without assumptions would sound most future-proof to me.

wddgit replied:
With the newest version I just pushed this part of the code is deleted.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47996/44679

cmsbuild commented:
Pull request #47996 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

wddgit commented:
I was manually running a test with FWCore/Framework/test/test_module_delete_cfg.py then looking at the graphic file produced. In the code from before this PR this looked OK. With the original version of the PR, the edges were messed up. The edges had been shifted to incorrect vertexes. I spent some time trying to understand why and could find nothing in my changes that seemed like it could reasonably cause this. Documentation of boost graph does not explain how this is supposed to work and I failed trying to get it to work just randomly trying variations in the code, but I did trace the problem to the deletion of modules. If vertexes are not deleted from the boost graph, then the edges are fine. They only get messed up when the vertexes are deleted.

Then I tried a different approach. We already know how many modules exist at the point the graph is populated. There is no reason to create extra vertexes and then delete ones corresponding to deleted modules. I changed the code to just create the correct number in the first place. Deleting the modules is no longer necessary. This seems more resilient. I also fixed it so it will continue to work if the numbers of sources changes. The problems I was seeing with the edges getting connected to the wrong module vanished. I think this is better.

wddgit commented:
please test

wddgit commented:
Comparisons look good. There were only the usual failures related to issue #47071, the MessageLogger and the ErrorHarvester. Nothing related to changes in this PR.

fwyzard commented:
I see, thanks, the new approach makes sense.

Rather than an `std::map<unsigned int, std::size_t>` how about using an `std::vector<std::size_t>` for the mapping from module id to vertex id ? You could call `resize(pathsAndConsumes.largestModuleID() + 1, kInvalidVertex)` in `lookupInitializationComplete` to avoid the automatic resizing.

`kInvalidVertex` could be `std::numeric_limits<std::size_t>::max()` - the value shouldn't matter anyway, as the vertices that are not overriden with a value vertex id correspond to the deleted modules, and should never be accessed anyway.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-47996/44695

cmsbuild commented:
Pull request #47996 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

makortel commented:
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/47071

makortel commented:
+core

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @antoniovilela, @mandrenguyen, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
